app/lua/lnodemcu.c

```diff
@@ -0,0 +1,116 @@
+/*
+** $Id: ltm.c,v 2.8.1.1 2007/12/27 13:02:25 roberto Exp $
+** Tag methods
+** See Copyright Notice in lua.h
+*/
+
+
+#define lnodemcu_c
+#define LUA_CORE
+
+#include "lua.h"
+#include <string.h>
+
+#include "lobject.h"
+#include "lstate.h"
+#include "lauxlib.h"
+#include "lgc.h"
+#include "lstring.h"
+#include "ltable.h"
+#include "ltm.h"
+#include "lrotable.h"
+#include "platform.h"
+
+extern int debug_errorfb (lua_State *L);
+#if 0
+extern int pipe_create(lua_State *L);
+extern int pipe_read(lua_State *L);
+extern int pipe_unread(lua_State *L);
+extern int pipe_write(lua_State *L);
+#endif
+/*
+** Error Reporting Task.  We can't pass a string parameter to the error reporter
+** directly through the task interface the call is wrapped in a C closure with
+** the error string as an Upval and this is posted to call the Lua reporter.
+*/
+static int report_traceback (lua_State *L) {
+// **Temp**  lua_rawgeti(L, LUA_REGISTRYINDEX, G(L)->error_reporter);
+  lua_getglobal(L, "print");
+  lua_pushvalue(L, lua_upvalueindex(1));
+  lua_call(L, 1, 0);  /* Using an error handler would cause an infinite loop! */ 
+  return 0;
+}
+
+/*
+** Catch all error handler for CB calls.  This uses debug.traceback() to
+** generate a full Lua traceback.  
+*/
+int luaN_traceback (lua_State *L) {
+  if (lua_isstring(L, 1)) {
+    lua_pushlightfunction(L, &debug_errorfb);
+    lua_pushvalue(L, 1);                                /* pass error message */
+    lua_pushinteger(L, 2);                /* skip this function and traceback */
+    lua_call(L, 2, 1);      /* call debug.traceback and return it as a string */
+    lua_pushcclosure(L, report_traceback, 1);     /* report with str as upval */
+    luaN_posttask(L, LUA_TASK_HIGH);
+  } 
+  return 0;
+}
+
+/*
+** Use in CBs and other C functions to call a Lua function. This includes
+** an error handler which will catch any error and then post this to the 
+** registered reporter function as a separate follow-on task.
+*/
+int luaN_call (lua_State *L, int narg, int res, int doGC) { // [-narg, +0, v]
+  int status;
+  int base = lua_gettop(L) - narg;
+  lua_pushcfunction(L, luaN_traceback);
+  lua_insert(L, base);                                      /* put under args */
+  status = lua_pcall(L, narg, (res < 0 ? LUA_MULTRET : res), base);
```

Comment:
Q: Would we not be better off having `luaN_call()` take the potential error message off the stack here? I know from experience that it's something quite easily forgotten by the caller since normally things don't fail and there's nothing left on the stack. Alternatively, push a nil on success to force the caller to balance the stack (less clean, but would provide the error message to the caller).

Reply:
See general response below.  `luaN_call()` does take the error off the task (see `luaN_traceback()` and `report_traceback()`.  Ultimately the app programmer will have the option of establishing an error reporter.   If the library routine wants to manage the error itself then it will already be using a `lua_pcall()`.  IIRC there is only one instance of this in the existing modules.  

This is designed to be a one-line swap-out in such library CB executions, and they don't have or have never tested error paths, since the current runtime simply throws a PANIC error and reboots the system.  Can I suggest if we want to fine tune this functionality then let us raise a separate specific issue.

Reply:
Agreed that we can tweak specifics later. I just want to make sure we don't accidentally leave the result of an errored pcall on the stack and thus unbalancing it.